Homework2/Solution

Inzilbeth opened the pull request.

Inzilbeth committed:
Added solution basis

Inzilbeth committed:
Solution renamed

Inzilbeth committed:
Create LazyMultiThreaded.cs

Inzilbeth committed:
Renamed solution folder

Inzilbeth committed:
Main functional implemented & documented

Inzilbeth committed:
Added tests, fixed some behavior

- Made LayFactory public, removed "Lazy" at the end of it's methods
- Overrided Equals and GetHashCode in LazySingleThreaded & LazyMultiThreaded
- Renamed variables in one style throughout the codebase
- Added tests for LazyFactory, LazySingleThreaded & LazyMultiThreaded

yurii-litvinov commented on Homework2/Task1/Task1Library/ILazy.cs:
```diff
@@ -0,0 +1,13 @@
+/// Interface for lazy instantionation of an object.
+/// </summary>
+/// <typeparam name="T">Type of the object to be instantinated.</typeparam>
+public interface ILazy<T>
```
Странно, что интерфейс вне какого-либо пространства имён, а все остальные классы в Task1Library

yurii-litvinov commented on Homework2/Task1/Task1Library/LazyFactory.cs:
```diff
@@ -0,0 +1,32 @@
+        public static LazySingleThreaded<T> CreateSingleThreaded<T>(Func<T> supplier)
+        {
+            return new LazySingleThreaded<T>(supplier);
+        }
```
Можно было через =>

yurii-litvinov commented on Homework2/Task1/Task1Library/LazyMultiThreaded.cs:
```diff
@@ -0,0 +1,77 @@
+        /// <summary>
+        /// Checks if an instance is already initialized.
+        /// </summary>
+        public bool IsInitialized { get => isInitialized;}
```
Это то же самое, что просто `public bool IsInitialized => isInitialized;`

yurii-litvinov commented on Homework2/Task1/Task1Library/LazyMultiThreaded.cs:
```diff
@@ -0,0 +1,77 @@
+
+                return instance.Equals(that.instance) &&
+                    isInitialized.Equals(that.isInitialized) &&
+                    supplier.Equals(that.supplier);
```
Ох какой потоконебезопасный Equals. Пока Вы сравниваете, другой поток может Вам поля менять

yurii-litvinov commented on Homework2/Task1/Task1Library/LazyMultiThreaded.cs:
```diff
@@ -0,0 +1,77 @@
+                isInitialized.GetHashCode() ^
+                supplier.GetHashCode();
+
+            return hashCode;
```
И тут, мы можем увидеть хеш-код объекта, которого никогда в природе не существовало, потому что прямо когда мы его считали, объект изменился. Конечно, можно сказать, что клиент сам дурак, но надо хотя бы в комментарии пометить, что этот метод не потокобезопасный ни разу.

yurii-litvinov commented on Homework2/Task1/Task1Tests/LazyTests.cs:
```diff
@@ -0,0 +1,122 @@
+        {
+            var testObject = LazyFactory.CreateMultiThreaded<int>(() => 1);
+            Assert.AreEqual(1, testObject.Get());
+        }
```
Кажется, что сие есть копипаст однопоточного случая. Потому что в этих тестах семантика обеих вариантов абсолютно одинакова, почему бы не тестировать их одним набором

yurii-litvinov reviewed (changes requested):
В целом как-то так, но можно ещё поправить.

Inzilbeth committed:
Fixes

- Added commentary for Equals and GetHashCode methods in both ILazy implementations
- Improved test structure

yurii-litvinov reviewed (approved):
Всё отлично, зачтена
